core/src/main/java/io/grpc/InternalLogId.java

```diff
@@ -25,38 +29,86 @@
  */
 @Internal
 public final class InternalLogId {
+
   private static final AtomicLong idAlloc = new AtomicLong();
 
   /**
-   * @param tag a loggable tag associated with this tag. The ID that is allocated is guaranteed
-   *            to be unique and increasing, irrespective of the tag.
+   * @param type the "Type" to be used when logging this id.   The short name of this class will be
+   *     used, or else a default if the class is anonymous.
+   * @param details a short, human readable string that describes the object the id is attached to.
+   *     Typically this will be an address or target.
+   */
+  public static InternalLogId allocate(Class<?> type, @Nullable String details) {
+    return allocate(getClassName(type), details);
+  }
+
+  /**
+   * @param typeName the "Type" to be used when logging this id.
+   * @param details a short, human readable string that describes the object the id is attached to.
+   *     Typically this will be an address or target.
    */
-  public static InternalLogId allocate(String tag) {
-    return new InternalLogId(tag, getNextId());
+  public static InternalLogId allocate(String typeName, @Nullable String details) {
+    return new InternalLogId(typeName, details, getNextId());
   }
 
   static long getNextId() {
     return idAlloc.incrementAndGet();
   }
 
-  private final String tag;
+  private final String typeName;
+  private final @Nullable String details;
   private final long id;
 
-  protected InternalLogId(String tag, long id) {
-    this.tag = tag;
+  InternalLogId(String typeName, String details, long id) {
+    checkNotNull(typeName, typeName);
+    checkArgument(!typeName.isEmpty(), "empty type");
+    this.typeName = typeName;
+    this.details = details;
     this.id = id;
   }
 
-  public long getId() {
-    return id;
+  public String getTypeName() {
+    return typeName;
+  }
+
+  @Nullable
+  public String getDetails() {
+    return details;
   }
 
-  public String getTag() {
-    return tag;
+  public long getId() {
+    return id;
   }
 
   @Override
   public String toString() {
-    return tag + "-" + id;
+    return formatName(typeName, details, id);
+  }
+
+  private static String formatName(String typeName, @Nullable String details, long id) {
+    StringBuilder sb = new StringBuilder();
+    sb.append(shortName(typeName, id));
+    if (details != null) {
+      sb.append(' ');
+      sb.append(details);
+    }
+    return sb.toString();
+  }
+
+  private static String getClassName(Class<?> type) {
+    String className = checkNotNull(type, "type").getSimpleName();
+    if (!className.isEmpty()) {
+      return className;
+    }
+    // + 1 removes the separating '.'
+    return type.getName().substring(type.getPackage().getName().length() + 1);
+  }
+
+  public String shortName() {
+    return shortName(typeName, id);
+  }
+
+  private static String shortName(String typeName, long id) {
```

Comment:
Why do you need this static method that are only called from `shortName()`?

Reply:
Fixed.